keps/sig-cli/kubectl-apply-prune-beta.md

```diff
@@ -0,0 +1,141 @@
+---
+title: Kubectl Apply Prune Enhancement
+authors:
+  - "@Liujingfang1"
+owning-sig: sig-cli
+participating-sigs:
+  - sig-cli
+reviewers:
+  - "@pwittrock"
+  - "@seans3"
+  - "@soltysh"
+approvers:
+  - "@pwittrock"
+  - "@seans3"
+  - "@soltysh"
+editor: "Liujingfang1"
+creation-date: 2019-02-04
+last-updated: 2019-02-04
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Kubectl Apply Prune Enhancement
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Challenges](#challenges)
+    * [Root Object](#root-object)
+    * [Child Objects](#child-objects)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks](#drawbacks)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+Link to tracking issue: [kubernetes/kubectl#572](https://github.com/kubernetes/kubectl/issues/572)
+
+## Summary
+
+`kubectl apply --prune` has been in Alpha for more than a year. This design proposes a better approach of handling `--prune` and move it to Beta. This feature targets 1.15 release.
+
+## Motivation
+
+`kubectl apply` operates an aggregation of Kubernetes Objects. While `kubectl apply` is widely used in different workflows, users expect it to be able to sync the objects on the cluster with configuration saved in files, directories or repos. That means, when `kubectl apply` is run on those configurations, the corresponding objects are expected to be correctly created, updated and pruned. 
+
+  Currently `kubectl apply --prune` is still in Alpha. It is based on label matching. Users need to be really careful to use this command, so that they don't delete objects unintentionally. There is a list of [issues]([kubernetes/kubectl#572]) observed by users. The most noticeable issue happens when changing directories or working with directories with the same labels. When two directories share the same label. `kubectl apply --prune` from one directory will delete objects from the other directory.
```

Comment:
User error can be challenging to avoid. It is hard to differentiate whether the error is really an error or a user intent. IMO, we should at least express it clearly how `prune` works for some popular user errors.